### PR TITLE
8269957: facilitate alternate impls of NameTable and Name

### PR DESCRIPTION
**Pull request by archiecobbs**
The `Name.Table` class is used by the compiler to hold unique instances of strings as `Name` objects.

In theory the `Name` superclass supports alternate implementations beyond the two existing implementations (`SharedNameTable` and `UnsharedNameTable`), but its current design presumes that strings are stored as UTF-8 byte arrays, which discourages other approaches.

The goal of this PR is to refactor things to allow for more flexibility in alternate `Name` implementations.

As a simple test case of this idea, it should be relatively simple to implement a `Name.Table` that stores `String`s in a hash table. This patch includes such an example in the new class `StringNameTable`, which can be enabled via the `-XDuseStringTable=true` command line flag.  A simple performance test with this class enabled ([JavacNameTable.java.txt](https://github.com/openjdk/jdk/files/11602852/JavacNameTable.java.txt)) shows a 17% speedup.

Changes:
* Remove all byte-oriented methods from the `Name` and `Name.Table` API's, except for those that import/export Modified UTF-8.
* Change the semantics of `Name.subName()` so the offset is a character offset, not a byte offset.
* Consolidate the common UTF-8 machinery of `SharedNameTable` and `UnsharedNameTable` into a new common superclass `Utf8NameTable`.
* Rename `Name.lastIndexOf()` -> `Name.lastIndexOfAscii()` to more accurately reflect its expected behavior.
* Add new `StringNameTable` implementation.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8269957](https://bugs.openjdk.org/browse/JDK-8269957): facilitate alternate impls of NameTable and Name (**Bug** - P3)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13282/head:pull/13282` \
`$ git checkout pull/13282`

Update a local copy of the PR: \
`$ git checkout pull/13282` \
`$ git pull https://git.openjdk.org/jdk.git pull/13282/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13282`

View PR using the GUI difftool: \
`$ git pr show -t 13282`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13282.diff">https://git.openjdk.org/jdk/pull/13282.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13282#issuecomment-1496468340)

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
Is it worth adding `StringNameTable` into the PR?

**Comment by jonathan-gibbons** [comment]
Is it worth putting static factory methods on `Name.Table` ?

**Comment by archiecobbs** [comment]
> Is it worth adding `StringNameTable` into the PR?

I was going to ask you the same question :) 

It's probably the right thing to do unless there's some rule against including code that's not normally going to be used.

We can add another secret flag to enable it. Currently we have `-XDuseUnsharedTable=true` which enables `UnsharedNameTable`, so we could add `-XDuseStringTable` to enable `StringNameTable`.

Added in 23bb34511cf.

> Is it worth putting static factory methods on `Name.Table` ?

Good idea.. but I think maybe they make more sense being added to `Names` rather than `Name.Table` since a `Names` object is a required parameter.

Added in c9cf3032819.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj == null || obj.getClass() != getClass())
+                return false;
+            final NameImpl that = (NameImpl)obj;
+            return that.table == table && that.string.equals(string);
```
Comment (ExE-Boss) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java:111]: This can use `instanceof <pattern>` to match `Utf8NameTable.NameImpl::equals`:
```suggestion
            return (obj instanceof NameImpl name)
                && table == name.table
                && string.equals(name.string);
```

Comment (archiecobbs) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java]: Thanks. I agree the two implementations should be consistent. However you also made me realize that some common logic (e.g., type and same table checks) can be pushed up into the superclass. Also I think we should keep the initial same object check though for performance.

Refactored in c5a2546e8a5.

**Comment by jonathan-gibbons** [comment]
> It's probably the right thing to do unless there's some rule against including code that's not normally going to be used.
> 

Well, while totally dead code is a bad thing, there is potential to explore the use of `StringNameTable` in both javadoc, and maybe even for javac.

**Comment by archiecobbs** [comment]
I figured out how to run performance tests and ran tests using `SharedNameTable`, `UnsharedNameTable`, and the new `StringNameTable`.

The latter is about 17% faster:
```
Benchmark                         Mode  Cnt   Score   Error  Units
JavacNameTable.testSharedTable    avgt   25  52.601 ± 1.025  ms/op
JavacNameTable.testStringTable    avgt   25  43.331 ± 1.165  ms/op
JavacNameTable.testUnsharedTable  avgt   25  57.687 ± 1.728  ms/op
```
Here is the test I ran: [JavacNameTable.java](https://github.com/openjdk/jdk/files/11602852/JavacNameTable.java.txt)

**Comment by archiecobbs** [comment]
pingbot

**Comment by jonathan-gibbons** [comment]
Verified javadoc tests all pass if StringNameTable is on by default in javadoc

**Comment by jonathan-gibbons** [comment]
Verified API documentation is unchanged if `StringnameTable` is on by default in javadoc

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+// Factory
+
+    public static StringNameTable create(Names names) {
+        return new StringNameTable(names);
+    }
+
+// Constructors
+
+    public StringNameTable(Names names) {
+        this(names, 8000);
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java:51]: did you mean: 0x8000?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/Utf8NameTable.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 1999, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/Utf8NameTable.java:2]: nit: copyright year should only be 2023

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java (hunk 2 above)
Comment (archiecobbs) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java]: Yes, thanks... was trying to cut & paste from `SharedNameTable()` but dropped the `0x` somehow. Fixed in 3c6d9145219.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/Utf8NameTable.java (hunk 3 above)
Comment (archiecobbs) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/Utf8NameTable.java]: Thanks, fixed in 193d06662dd.

**Comment by jonathan-gibbons** [comment]
CI job for all tier1 tests, with `StringNameTable` enabled by default, passed.

**Review by jonathan-gibbons — APPROVED** [review]
The code passes javadoc tests and :tier1 tests with StringNameTable enabled.

I'll approve, but recommend we get Vicente's approval, as a javac dev, as well.

**Comment by archiecobbs** [comment]
> I'll approve, but recommend we get Vicente's approval, as a javac dev, as well.

Thanks @jonathan-gibbons. I agree and will bump the required reviewer count.

/reviewers 2 reviewer

**Review by vicente-romero-oracle — APPROVED** [review]
looks good to me too, thanks for fixing this!

**Pull request closed without merging**